reconcile/openshift_prometheusrules.py

```diff
@@ -0,0 +1,273 @@
+    params['latency_rates'] = []  # sloRules of latency_rate type
+    params['all_rules'] = []      # contains the name of each rule.
+
+    # We have checked that rule names are unique by category, but not as whole
+    # we will add a suffix to each rule to make sure they're unique
+    for r in pp['sloRules']:
+        if r['kind'] == 'http_rate':
+            params['http_rates'].append(r)
+            params['all_rules'].extend(
+                ['%s__http_rates.rateRules' % r['name'],
+                 '%s__http_rates.errorRateRules' % r['name']])
+        else:
+            params['latency_rates'].append(r)
+            params['all_rules'].append('%s__latency_rates.rules' % r['name'])
+
+    for c in ['volume', 'latency', 'errors', 'availability']:
+        params['all_rules'].extend(['%s__%s.rules' % (r['name'], c)
+                                    for r in pp[c]])
+        params[c] = pp[c]
+
+    params['labels'] = pp['labels']
+    params['component'] = pp['component']
+    params['namespace'] = pp['namespace']['name']
+    params['prometheus_labels'] = pp['prometheusLabels']
+
+    return params
+
+
+def fetch_desired_state(performance_parameters, ri):
+    for pp in performance_parameters['performance_parameters_v1']:
+        if pp['namespace']['cluster']['observabilityNamespace'] is None:
+            logging.error('No observability namespace for %s' % pp['name'])
+            continue
+
+        try:
+            errors = check_data_consistency(pp)
+            if len(errors) > 0:
+                logging.error(
+                    'Data inconsistent for %s. Errors detected: %s' % (
+                        pp['name'], errors))
+                continue
+
+            params = build_template_params(pp)
+            rules_resource = generate_resource(SLO_RULES, params)
+        except Exception as e:
+            logging.error('Error building resource for %s: %s' % (
+                pp['name'], e))
+            logging.debug(traceback.format_exc())
+            continue
+
+        cluster_observability_namespace = \
+            pp['namespace']['cluster']['observabilityNamespace']
+        ri.add_desired(
+            cluster_observability_namespace['cluster']['name'],
+            cluster_observability_namespace['name'],
+            'PrometheusRule',
+            rules_resource.body['metadata']['name'],
+            rules_resource)
+
+
+@defer
+def run(dry_run=False, thread_pool_size=10, internal=None,
+        use_jump_host=True, defer=None):
+    gqlapi = gql.get_api()
+    performance_parameters = gqlapi.query(PERFORMANCE_PARAMETERS_QUERY)
+    observability_namespaces = [
+        pp['namespace']['cluster']['observabilityNamespace']
+        for pp in performance_parameters['performance_parameters_v1']
+        if pp['namespace']['cluster']['observabilityNamespace'] is not None]
+
+    if len(observability_namespaces) == 0:
+        logging.error('No observability namespace matching')
+        return
```

Comment:
So you would make the whole integration fail if one performance parameter file is bad? I'm trying hard to avoid that

If we can change it to be mandatory it would be better

Reply:
Unless there's something I don't understand here, your code does exactly the same than the current code 😅